kimchi/src/circuits/polynomials/keccak/circuitgates.rs

```diff
@@ -0,0 +1,211 @@
+//~ | y = 1 |   55  |  276  |   36  |  300  |    6  |
+//~ | y = 0 |   28  |   91  |    0  |    1  |  190  |
+//~ | y = 4 |  120  |   78  |  210  |   66  |  253  |
+//~ | y = 3 |   21  |  136  |  105  |   45  |   15  |
+//~
+//~ ##### Design Approach:
+//~
+//~ The atomic operations are XOR, ROT, NOT, AND. In the sections below, we will describe
+//~ the gates for these operations. Below are some common approaches followed in their design.
+//~
+//~ To fit within 15 wires, we first decompose each word into its lower and upper 32-bit
+//~ components. A gate for an atomic operation works with those 32-bit components at a time.
+//~
+//~ Before we describe the specific gate design approaches, below are some constraints in the
+//~ Kimchi framework that dictated those approaches.
+//~ * only 4 lookups per row
+//~ * only first 7 columns are available to the permutation polynomial
+//~
+//~ ##### 32-bit decomposition gate
+//~
+//~ This is a basic operation that is typically done for 64-bit initial state and
+//~ intermediate values.
+//~
+//~ Let `inp` be a 64-bit word. The constraint is: `in` $= 2^32 \cdot$ `in_hi` $+$ `in_lo`.
+//~ It takes 3 cells for values `in`, `in_hi`, `in_lo`. We have not yet placed them w.r.t
+//~ other rows of the Keccak computation; the only requirement is that all these cells be
+//~ within the first 7 columns for permutation equation accessibility.
+//
+//~ ##### XOR gate
+//~
+//~ First we consider a XOR gate that checks that a 32-bit word `out` is the XOR of `in1` and `in2`.
+//~ This gate will use 2 rows, with a `Xor` row followed by a `Zero` row.
+//~
+//~ | Gates |          `Xor`   |          `Zero`  |
+//~ | ----- | ---------------- | ---------------- |
+//~ | Rows  |           0      |           1      |
+//~ | Cols  |                  |                  |
+//~ |     0 | copy     `in1`   | copy     `out`   |
+//~ |     1 |                  | copy     `in2`   |
+//~ |     2 |                  |                  |
+//~ |     3 | plookup0 `in2_0` | plookup4 `in2_4` |
+//~ |     4 | plookup1 `in2_1` | plookup5 `in2_5` |
+//~ |     5 | plookup2 `in2_2` | plookup6 `in2_6` |
+//~ |     6 | plookup3 `in2_3` | plookup8 `in2_7` |
+//~ |     7 | plookup0 `in1_0` | plookup4 `in1_4` |
+//~ |     8 | plookup1 `in1_1` | plookup5 `in1_5` |
+//~ |     9 | plookup2 `in1_2` | plookup6 `in1_6` |
+//~ |    10 | plookup3 `in1_3` | plookup7 `in1_7` |
+//~ |    11 | plookup0 `out_0` | plookup4 `out_4` |
+//~ |    12 | plookup1 `out_1` | plookup5 `out_5` |
+//~ |    13 | plookup2 `out_2` | plookup6 `out_6` |
+//~ |    14 | plookup3 `out_3` | plookup7 `out_7` |
+//~
+//~ Now we apply this gate twice to obtain a XOR gadget for 64-bit words by halving:
+//~
+//~ Consider the following operations:
+//~ * `out_lo` $=$ `in1_lo` $\oplus$ `in2_lo` and
+//~ * `out_hi` $=$ `in1_hi` $\oplus$ `in2_hi`,
+//~ where each element is 32 bits long.
+//~
+//~ | Gates |    `Xor` |   `Zero` |    `Xor` |   `Zero` |
```

Comment:
It feels rather tempting to have a 'chainable' gate here, along the lines of

```
        0   1   2   3   4   5   6   7   8   9   10  11  12  13  14
XOR     x   y   z   x_0 y_0 z_0 x_1 y_1 z_1 x_2 y_2 z_2 x_3 y_3 z_3
        x'  y'  z'
```

where you have x, y, z arbitrary-size values, and constrain
```
x_0 XOR y_0 = z_0
x_1 XOR y_1 = z_1
x_2 XOR y_2 = z_2
x_3 XOR y_3 = z_3
x = x_0 + 2^4 * x_1 + 2^(2*4) * x_2 + 2^(3*4) * x_3 + 2^(8*4) * x'
y = y_0 + 2^4 * y_1 + 2^(2*4) * y_2 + 2^(3*4) * y_3 + 2^(8*4) * y'
z = z_0 + 2^4 * z_1 + 2^(2*4) * z_2 + 2^(3*4) * z_3 + 2^(8*4) * z'
```

(that is, x', y', and z' are the remaining high bits of x, y, and z respectively)

With this, you would witness x, y, z, claiming that `x XOR y = z` in 64-bits, then run the gate twice and constrain the final outputs to be 0. (Likewise, you would run it 4 times for an 128-bit XOR.)

This also appears to use 3 rows (1 partial, could be reused for a generic) rather than 4.

Reply:
I really really like this approach! Specially because it does save in number of rows. Meaning, the original proposal needs 4 rows and 1,5 decomposition gates, for a total of 5.5 rows. Whereas here as you say, it is 4,5 or 5, depending on whether we reuse the final row of three leading zeros for something else (or we create a XORfinal gate that does not need to access Next, but costs one gate type more). 

Nonetheless, I think the offset for the implicit decomposition constrain would not be a fixed value but a power of 2^16 each time. Meaning, for each x y and z:

``` 
XOR1: x = x_0 + 2^4 * x_1 + 2^8 * x_2 + 2^12 * x_3 + 2^16 * x' (where x' has 48 bits)
XOR2: x' = x'_0 + 2^4 * x'_1 + 2^8 * x'_2 + 2^12 * x'_3 + 2^32 * x'' (where x'' has 32 bits) 
XOR3: x'' = x''_0 + 2^4 * x''_1 + 2^8 * x''_2 + 2^12 * x''_3 + 2^32 * x''' (where x''' has 16 bits) 
XOR4: x''' = x'''_0 + 2^4 * x'''_1 + 2^8 * x'''_2 + 2^12 * x'''_3 + 2^32 * x'''' (where x'''' is zero) 
```
Thus, I guess the offset could be a coefficient of the xor gate. What do you think?

Reply:
I am pasting a chat with Matthew so everyone can read:

This is a different question, but if in the future we had an 8-bit XOR table instead, we would just need 2 rows per xor, period. No need for the 0,5 final one.

Just to have some numbers, each keccak gadget will need about 3 thousand 64-bit XORs. And going from the current 4-bit XOR table (256 rows) to an 8-bit one would mean 65k more rows. The new design will then need about 15k xor rows for 1 keccak, and the other one 6k. So as soon as we needed 8 keccaks in one single circuit, then it would make sense to have an 8-bit xor table. Right?

Oh, this reminds me of a question about plookup tables. Since XOR is commutative, do we have any means to check whether (a,b) belongs to the table, or (b,a)? I am asking because if we had, then we could halve the rows of the 8-bit xor table, and thus it could make sense to have it as soon as we needed as little as 4 keccaks in one circuit.

Reply:
> Since XOR is commutative, do we have any means to check whether (a,b) belongs to the table, or (b,a)?

Not that I'm aware of -- the entries in the table are encoded as a single value that represents the *ordered* tuple of values, so changing the order gives an entirely different value. I'm interested to hear any suggestions if you think we can tweak the plookup argument to allow for this though!